### PR TITLE
Implement on demand approach to keycloak realm provisioning for multi tenant Rhoam in user sso.

### PR DESCRIPTION
**Pull request by briangallagher**
# Issue link
Implement on demand approach to keycloak realm provisioning for multi tenant Rhoam in user sso. 

# Verify
Install MT RHOAM from this branch
Install IDP:
`NUM': NUM_REGULAR_USER=20 NUM_ADMIN=0 DEDICATED_ADMIN_PASSWORD=Password1 PASSWORD=Password1 REALM_DISPLAY_NAME="devsandbox" REALM="devsandbox" ./scripts/setup-sso-idp.sh` 
Log into OpenShift with 4 users test-user01, test-user02, test-user03, test-user04
This will create user and identity CRs

Wait for multi-tenant users to reconcile
Log in as admin to User SSO and verify that the Staging Realm exists with 4 users

Create projects and related privileges for each user
`oc new-project test-user01`
`oc adm policy add-role-to-user admin test-user01 -n test-user01-dev`
repeat for users 2

Log into OpenShift as test-user01 
Go to the test-user01-dev project and then to the Dashboard
From the Launcher Section, login to the staging Realm by clicking on 'Provision SSO' 
Repeat for test-user02

Wait for a full reconcile loop

Log in as admin to User SSO and verify that the Staging Realm exists with **2 users**: 3 and 4
Confirm that 2 new realms exist, test-user01 and test-user02

Log into OpenShift as test-user01 
Go to the test-user01-dev project and then to the Dashboard
From the Launcher Section, login to the staging Realm by clicking on 'API Management SSO' 
Verify that the user is logged into the test-user01 realm with full admin access
Repeat for test-user02

Delete 4 user crs and identity crs

Wait for a full reconcile loop

Confirm that there are no users in the staging realm and no tenant realms
Only a staging and master realm

### CODE REVIEW

File: pkg/products/rhssouser/reconciler_multitenant.go [hunk 1]
```diff
@@ -0,0 +1,710 @@
...
+		_, err = r.createTenantKCUser(ctx, serverClient, user)
+		if err != nil {
+			return err
+		}
+		if err = r.removeUserFromStagingRealm(ctx, serverClient, user); err != nil {
+			return err
+		}
+	}
+
+	// remove user from staging realm
```
Comment (MStokluska) [inline pkg/products/rhssouser/reconciler_multitenant.go:124]: Hey, something small but I think this should be //deleting tenant or so, if I understand correctly we are removing user from staging realm at 119.

File: pkg/products/rhssouser/reconciler_multitenant.go [hunk 2]
```diff
@@ -0,0 +1,710 @@
...
+	kcClient, err := r.KeycloakClientFactory.AuthenticatedClient(*kc)
+	if err != nil {
+		return nil, err
+	}
+	return kcClient.ListOfActivesUsersPerRealm(StagingRealmName, "", 1000)
+}
+
+func (r *Reconciler) reconcileStagingRealm(ctx context.Context, serverClient k8sclient.Client, realms keycloak.KeycloakRealmList) error {
+
+	r.Log.Info("Creating staging realm")
```
Comment (MStokluska) [inline pkg/products/rhssouser/reconciler_multitenant.go:276]: I would update to "Reconciling staging realm"

File: pkg/products/rhssouser/reconciler_multitenant.go [hunk 3]
```diff
@@ -0,0 +1,710 @@
...
+			Name:      StagingRealmName,
+			Namespace: r.Config.GetNamespace(),
+		},
+	}
+
+	or, err := controllerutil.CreateOrUpdate(ctx, serverClient, kcr, func() error {
+		kcr.Spec.RealmOverrides = []*keycloak.RedirectorIdentityProviderOverride{
+			{
+				IdentityProvider: idpAlias,
+				ForFlow:          "browser",
```
Comment (MStokluska) [inline pkg/products/rhssouser/reconciler_multitenant.go:288]: Just marking it as a reminder to ask about this on a sync call with you!

File: pkg/products/rhssouser/reconciler_multitenant.go [hunk 4]
```diff
@@ -0,0 +1,710 @@
...
+	return nil, nil
+}
+
+func (r *Reconciler) getUserIdLoginEvents(ctx context.Context, serverClient k8sclient.Client, kc *keycloak.Keycloak) ([]kccommon.Users, error) {
+
+	kcClient, err := r.KeycloakClientFactory.AuthenticatedClient(*kc)
+	if err != nil {
+		return nil, err
+	}
+	return kcClient.ListOfActivesUsersPerRealm(StagingRealmName, "", 1000)
```
Comment (jjaferson) [inline pkg/products/rhssouser/reconciler_multitenant.go:271]: are we defaulting to 1000 realms even though we know keycloak won't be able to create it?

**Review by jjaferson — COMMENTED** [review]
Code looks fine, just left one question there

**Comment by MStokluska** [comment]
Reviewed code and run the verification steps - great job! Code looks good to me (just two small questions regarding some comments).
Also, I've run the verification steps and everything works as expected.
Good job!

/lgtm

File: vendor/github.com/integr8ly/keycloak-client/pkg/common/client.go [hunk 5]
```diff
@@ -952,6 +952,41 @@ func (c *Client) CreateGroupRealmRole(role *v1alpha1.KeycloakUserRole, realmName
...
+		"eventsEnabled":     "true",
+		"eventsListeners":   eventsListeners,
+		"enabledEventTypes": enabledEventTypes,
+	}
+	path := fmt.Sprintf("realms/%s/events/config", realmName)
+
+	return c.update(add, path, "events-config")
+}
+
+func (c *Client) ListOfActivesUsersPerRealm(realmName, dateFrom string, max int) ([]Users, error) {
```
Comment (MStokluska) [inline vendor/github.com/integr8ly/keycloak-client/pkg/common/client.go:966]: Just spotted a typo here?

File: pkg/products/rhssouser/reconciler_multitenant.go (hunk 4 above)
Comment (briangallagher) [reply pkg/products/rhssouser/reconciler_multitenant.go:271]: 1000 login events. That should be ok?

**Comment by MStokluska** [comment]
/lgtm again due to unit test update.
On previous run, rhoam e2e was a success and e2e (rhmi) failed on the incorrect version of Apicurito (not related to this PR)

**Pull request merged**